File: nbrk_currency_rates/src/main/java/com/nbrk/rates/activity/MainActivity.java
package com.nbrk.rates.activity;

import android.app.DatePickerDialog;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.DatePicker;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.nbrk.rates.CurrencyRatesService;
import com.nbrk.rates.DatePickerDialogFragment;
import com.nbrk.rates.R;
import com.nbrk.rates.RatesAdapter;
import com.nbrk.rates.model.CurrencyRatesItem;
import de.greenrobot.event.EventBus;
import de.greenrobot.event.Subscribe;
import de.greenrobot.event.ThreadMode;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class MainActivity extends ActionBarActivity implements DatePickerDialog.OnDateSetListener,
        SwipeRefreshLayout.OnRefreshListener {

    ListView list;
    TextView title;
    //MoPubView adview;

    String sTitle;

    String loading;

    Calendar date;
    SimpleDateFormat sdf = new SimpleDateFormat("dd.MM.yyyy");
    SimpleDateFormat sdf2 = new SimpleDateFormat("dd MMMM yyyy (HH:mm)");
    List<CurrencyRatesItem> currencyRatesItemList = new ArrayList<>();
    RatesAdapter ratesAdapter;
    SwipeRefreshLayout mSwipeRefreshLayout;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Toolbar toolbar = (Toolbar) findViewById(R.id.my_awesome_toolbar);
        setSupportActionBar(toolbar);

        list = (ListView) findViewById(R.id.lv_currency_rates);
        title = (TextView) findViewById(R.id.title);
        mSwipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.refresh);
        mSwipeRefreshLayout.setOnRefreshListener(this);

        sTitle = getResources().getString(R.string.title);
        loading = getResources().getString(R.string.loading);

        // set current date
        date = Calendar.getInstance();

        loadRatesAsync(sdf.format(date.getTime()), false);

        AdView mAdView = (AdView) findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice(AdRequest.DEVICE_ID_EMULATOR)
                .build();
        mAdView.loadAd(adRequest);
    }

    public boolean isOnline() {
        ConnectivityManager cm = (ConnectivityManager)getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        return activeNetwork != null && activeNetwork.isConnectedOrConnecting();
    }

    void loadRatesAsync(String date, boolean refresh) {
        //Check connection
        if (isOnline()) {
            //Call service
            Intent intent = new Intent(getApplicationContext(), CurrencyRatesService.class);
            intent.putExtra("date", date);
            intent.putExtra("refresh", refresh);
            startService(intent);
        } else {
            Toast.makeText(this, R.string.no_network_connection, Toast.LENGTH_LONG).show();
        }
    }

    //public void onEventMainThread(List<CurrencyRatesItem> currencyRatesItemList) {
    @Subscribe(threadMode = ThreadMode.MainThread)
    public void handle(List<CurrencyRatesItem> currencyRatesItemList) {
        mSwipeRefreshLayout.setRefreshing(false);
        title.setText(sTitle + " " + sdf2.format(date.getTime()));
        this.currencyRatesItemList = currencyRatesItemList;
        if (ratesAdapter == null) {
            ratesAdapter = new RatesAdapter(getApplicationContext(), currencyRatesItemList);
            list.setAdapter(ratesAdapter);
        }
        ratesAdapter.setCurrencyRatesList(currencyRatesItemList);
        ratesAdapter.notifyDataSetChanged();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_refresh:
                date = Calendar.getInstance();
                loadRatesAsync(sdf.format(date.getTime()), true);
                return true;
            case R.id.menu_date:
                showDatePickerDialog();
                return true;
            case R.id.menu_settings:
                Intent intent = new Intent(this, SettingsActivity.class);
                startActivity(intent);
                return true;
            case R.id.menu_converter:
                Intent converter = new Intent(this, ConverterActivity.class);
                EventBus.getDefault().postSticky(currencyRatesItemList);
                startActivity(converter);
                return true;
            case R.id.menu_chart:
                Intent chart = new Intent(this, ChartActivity.class);
                EventBus.getDefault().postSticky(currencyRatesItemList);
                startActivity(chart);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    void showDatePickerDialog() {
        Bundle bundle = new Bundle();
        bundle.putInt(DatePickerDialogFragment.YEAR, date.get(Calendar.YEAR));
        bundle.putInt(DatePickerDialogFragment.MONTH, date.get(Calendar.MONTH));
        bundle.putInt(DatePickerDialogFragment.DATE, date.get(Calendar.DAY_OF_MONTH));

        DatePickerDialogFragment datePicker = new DatePickerDialogFragment();

        datePicker.setArguments(bundle);
        datePicker.show(getFragmentManager(), "DatePickerDialogFragment");
    }

    @Override
    public void onDateSet(DatePicker datePicker, int year, int month, int day) {
        date.set(year, month, day);
        loadRatesAsync(sdf.format(date.getTime()), false);
    }

    @Override
    public void onPause() {
        EventBus.getDefault().unregister(this);
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        EventBus.getDefault().register(this);
        loadRatesAsync(sdf.format(date.getTime()), false);
    }

    @Override
    public void onRefresh() {
        // начинаем показывать прогресс
        mSwipeRefreshLayout.setRefreshing(true);
        date = Calendar.getInstance();
        loadRatesAsync(sdf.format(date.getTime()), true);
    }
}

File: nbrk_currency_rates/src/main/kotlin/com/nbrk/rates/home/view/activities/MainActivity.kt
package com.nbrk.rates.home.view.activities

import android.app.DatePickerDialog
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.widget.DatePicker
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdView
import com.nbrk.rates.App
import com.nbrk.rates.CurrencyRatesService
import com.nbrk.rates.R
import com.nbrk.rates.base.ToolbarManager
import com.nbrk.rates.chart.view.activities.ChartActivity
import com.nbrk.rates.converter.view.activities.ConverterActivity
import com.nbrk.rates.extensions.error
import com.nbrk.rates.extensions.toDateString
import com.nbrk.rates.entities.Rates
import com.nbrk.rates.home.presenter.MainPresenter
import com.nbrk.rates.home.view.adapters.RatesListAdapter
import com.nbrk.rates.settings.view.activities.SettingsActivity
import kotlinx.android.synthetic.main.activity_main.*
import nucleus.factory.RequiresPresenter
import nucleus.view.NucleusActivity
import org.jetbrains.anko.find
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.startService
import org.jetbrains.anko.support.v4.onRefresh
import org.jetbrains.anko.toast
import java.text.SimpleDateFormat
import java.util.*

@RequiresPresenter(MainPresenter::class)
class MainActivity : NucleusActivity<MainPresenter>(), ToolbarManager {

  var date = Calendar.getInstance()
  var sTitle = ""
  val sdf2 = SimpleDateFormat("dd MMM yyyy (HH:mm)")

  val onDateSet = fun(datePicker: DatePicker, year: Int, month: Int, day: Int) {
    date.set(year, month, day)
    loadRates()
  }

  override val toolbar by lazy { find<Toolbar>(R.id.toolbar) }

  override fun initToolbar() {
    toolbar.inflateMenu(R.menu.main_menu)
    toolbar.setOnMenuItemClickListener {
      when (it.itemId) {
        R.id.menu_refresh -> {
          //reset date to Today
          date = Calendar.getInstance()
          loadRates()
        }
        R.id.menu_date -> DatePickerDialog(this, onDateSet,
          date.get(Calendar.YEAR),
          date.get(Calendar.MONTH),
          date.get(Calendar.DAY_OF_MONTH)).show()
        R.id.menu_converter -> startActivity<ConverterActivity>()
        R.id.menu_chart -> startActivity<ChartActivity>()
        R.id.menu_settings -> startActivity<SettingsActivity>()
        else -> App.instance.toast("Unknown option")
      }
      true
    }
  }

  val ratesListAdapter by lazy { RatesListAdapter() }

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    initToolbar()
    toolbarTitle = getString(R.string.app_name)

    refresh.onRefresh {
      //reset date to Today
      date = Calendar.getInstance()
      loadRates()
    }

    sTitle = resources.getString(R.string.last_updated)

    rvRatesList.layoutManager = LinearLayoutManager(this)
    rvRatesList.adapter = ratesListAdapter

    val mAdView = findViewById(R.id.adView) as AdView
    val adRequest = AdRequest.Builder().addTestDevice(AdRequest.DEVICE_ID_EMULATOR).build()
    mAdView.loadAd(adRequest)

    if (savedInstanceState == null) {
      loadRates()
    }

  }

  fun loadRates() {
    //show progress bar
    refresh.isRefreshing = true
    //start service
    startService<CurrencyRatesService>(CurrencyRatesService.KEY_DATE to date.toDateString())
    //load from db
    presenter.loadRates(date.toDateString())
  }

  fun showRates(rates: Rates) {
    refresh.isRefreshing = false
    tvTitle.text = "$sTitle ${date.toDateString(sdf2)}"
    ratesListAdapter.setData(rates)
  }

  fun showError(error: Throwable) {
    refresh.isRefreshing = false
    error(error.message)
  }

}